[WIP] Segregated witness code

sipa opened the pull request.

jtimon committed:
Testchains: Don't check the genesis block

Rebased by Pieter Wuille.
Cleanup by Matt Corallo.

sipa committed:
Add segregated witness transaction serialization

Contains refactorings by Eric Lombrozo.
Contains fixup by Nicolas Dorier.

sipa committed:
Witness commitment validation

sipa committed:
Script validation logic for witnesses

sipa committed:
Negotiate witness fetching via 'havewitness'

sipa committed:
Enable SCRIPT_VERIFY_WITNESS for mempool transactions

sipa committed:
Activate script consensus rules in v5 blocks

sipa committed:
Signing tests

Commit:
[doc] Update release-process.md

Conflicts:
	doc/release-process.md

Github-Pull: #7465
Rebased-From: fa616c2fedd19d8e88f042abd5e99ac9595923df

sipa committed:
Make script validation observe input amounts

sipa committed:
Add signature version 1 with updated sighash

Includes simplifications by Eric Lombrozo.

sipa committed:
Only download blocks from witness peers after fork

sipa committed:
Witness script signing

morcos commented on src/main.cpp:
```diff
@@ -4179,25 +4298,25 @@ void static ProcessGetData(CNode* pfrom, const Consensus::Params& consensusParam
                     }
                 }
-                if (!pushed && inv.type == MSG_TX) {
+                if (!pushed && inv.type == MSG_TX || inv.type == MSG_WITNESS_TX) {
```
heh.  doesn't the && bind tighter than the ||?
I think you can just get rid of checking the inv.type, its checked in the containing `else if`

sipa replied:
Just noticed that myself.

morcos commented on src/main.cpp:
```diff
@@ -4179,25 +4298,25 @@ void static ProcessGetData(CNode* pfrom, const Consensus::Params& consensusParam
                         ss.reserve(1000);
                         ss << tx;
-                        pfrom->PushMessage(NetMsgType::TX, ss);
+                        pfrom->PushMessageWithFlag(inv.type == MSG_WITNESS_TX ? SERIALIZE_TRANSACTION_WITNESS : 0, NetMsgType::TX, ss);
```
The other thing I was trying to ask about is don't the two lines above this result in the tx first being serialized without the witness data, so then the `PushMessageWithFlag` wouldn't be able to serialize witness data because `ss` doesn't have it.

sipa replied:
About to fix.

sipa committed:
Add witness address RPCs (using P2SH)

Includes support for pushkeyhash wit v0 by Alex Morcos.

morcos commented on src/main.cpp:
```diff
@@ -1167,8 +1170,13 @@ bool AcceptToMemoryPoolWorker(CTxMemPool& pool, CValidationState &state, const C
         // This is done last to help prevent CPU exhaustion denial-of-service attacks.
-        if (!CheckInputs(tx, state, view, true, STANDARD_SCRIPT_VERIFY_FLAGS, true))
+        if (!CheckInputs(tx, state, view, true, STANDARD_SCRIPT_VERIFY_FLAGS, true)) {
+            if (CheckInputs(tx, state, view, true, STANDARD_SCRIPT_VERIFY_FLAGS & ~SCRIPT_VERIFY_WITNESS, true)) {
```
This line doesn't work because you still have CLEAN_STACK set and you can't have that set and not WITNESS.

sipa committed:
Create segnet

morcos committed:
Add command line options to loosen mempool acceptance rules

morcos committed:
Add rpc test for segwit

sipa committed:
Implementing base+wit/4 block size limit rule

Includes fixup by Suhas Daftuar.

jl2012 commented:
flag conflicts with BIP112:

SCRIPT_VERIFY_CHECKSEQUENCEVERIFY = (1U << 10),

morcos commented on src/script/sign.cpp:
```diff
@@ -79,62 +81,141 @@ static bool SignStep(const BaseSignatureCreator& creator, const CScript& scriptP
     txnouttype whichType;
-    if (!SignStep(creator, fromPubKey, scriptSig, whichType))
-        return false;
+    solved = SignStep(creator, script, result, whichType, 0);
```
I haven't thought through all the implications of why you might not want to short circuit here if solved is false.  However one issue is if you are trying to sign a p2sh script and don't have the ScriptID in your keystore, then result will be false and you'll core dump below in line 155.

sipa committed:
SigOp counting in witnesses

Includes fixup by Thomas Kerin.

rubensayshi commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1083,6 +1083,80 @@ UniValue addmultisigaddress(const UniValue& params, bool fHelp)
+{
+    if (fHelp || params.size() < 1 || params.size() > 1)
+    {
+        string msg = "createwitnessaddress \"address\"\n"
```
typo; addwitnessaddress

sipa replied:
Fixed.

CodeShark committed:
Removed ppszTypeName from protocol.cpp

luke-jr committed:
release-notes: Cover priority changes correctly, removing mentions of possible futures

theuni committed:
qt: Work (don't crash) with -segnet

CodeShark committed:
getdata enum issue fix

sdaftuar committed:
Increase MAX_PROTOCOL_MESSAGE_LENGTH

Witness blocks can be greater than 2MiB, but cannot be validly greater
than 4MB.

Commit:
[qt] Peertable: Increase SUBVERSION_COLUMN_WIDTH

Github-Pull: #7384
Rebased-From: faa9011d09d7429b97ec7595f9f77abf8ea770d3

sipa committed:
Add segnet seed nodes

sipa commented:
Added fixed seed nodes.

jl2012 committed:
Return witness data

afk11 committed:
bitcoinconsensus: add method that accepts amount, and return error when verify_script receives VERIFY_WITNESS flag

script_tests: always test bitcoinconsensus_verify_script_with_amount if VERIFY_WITNESS isn't set

Rename internal method + make it static

trim bitcoinconsensus_ prefix

Add SERIALIZE_TRANSACTION_WITNESS flag

luke-jr committed:
Rename permitrbf to replacebyfee

"permit" is currently used to configure transaction filtering, whereas replacement is more to do with the memory pool state than the transaction itself.

luke-jr committed:
Accept replacebyfee=opt-in for turning on opt-in RBF

Basic forward-compatibility with more flexible parameters like fss

luke-jr committed:
Simplify check for replacebyfee=opt-in

luke-jr committed:
release-notes: Update for permitrbf->replacebyfee rename

NicolasDorier committed:
Negotiate witness fetching via service bit 'NODE_WITNESS'

sipa committed:
Temporary: still send/receive 'havewitness' to/from old nodes

luke-jr committed:
Rename replacebyfee=opt-in to mempoolreplacement=fee

luke-jr committed:
release-notes: Update for replacebyfee->mempoolreplacement rename

laanwj committed:
Merge #7440: [0.12] Rename permitrbf to mempoolreplacement and provide minimal string-list forward compatibility

af9f564 release-notes: Update for replacebyfee->mempoolreplacement rename (Luke Dashjr)
4ad418b Rename replacebyfee=opt-in to mempoolreplacement=fee (Luke Dashjr)
b2287a7 release-notes: Update for permitrbf->replacebyfee rename (Luke Dashjr)
5f456a6 Simplify check for replacebyfee=opt-in (Luke Dashjr)
e8d19ab Accept replacebyfee=opt-in for turning on opt-in RBF (Luke Dashjr)
1205f87 Rename permitrbf to replacebyfee (Luke Dashjr)

laanwj committed:
doc: update release notes for rc3

laanwj committed:
qt: translations update pre-rc3

sdaftuar commented on src/main.cpp:
```diff
@@ -831,6 +847,11 @@ bool AcceptToMemoryPoolWorker(CTxMemPool& pool, CValidationState &state, const C
 
+    // Don't accept witness transactions before the final threshold passes
+    if (!GetBoolArg("-prematurewitness",false) && !tx.wit.IsNull() && !(chainActive.Tip()->nHeight + 1 >= Params().GetConsensus().SegWitHeight && IsSuperMajority(5, chainActive.Tip(), Params().GetConsensus().nMajorityRejectBlockOutdated, Params().GetConsensus()))) {
+        return state.DoS(0, false, REJECT_NONSTANDARD, "no-witness-yet");
```
This should set "corruption possible", or else someone could maliciously add bogus witness data to non-witness transactions and blind nodes to those transactions for an extended period of time.

mrbandrews committed:
[rpc-tests] Change solve() to use rehash

Github-Pull: #7468
Rebased-From: 7689041c03278a09c88a2bb78cd00217f6d4b1de

pastly committed:
Fix spelling: misbeha{b,v}ing

Github-Pull: #7469
Rebased-From: 0830552673e37142599de897e87510f2f9866e1e

laanwj committed:
rpc: Add WWW-Authenticate header to 401 response

A WWW-Authenticate header must be present in the 401
response to make clients know that they can authenticate,
and how.

    WWW-Authenticate: Basic realm="jsonrpc"

Fixes #7462.

Github-Pull: #7472
Rebased-From: 7c06fbd8f58058d77c3e9da841811201d2e45e92

laanwj committed:
doc: add PR authors to release notes

Take full names from github API if available, otherwise
github username.

instagibbs committed:
Changed getnetworkhps value to double to avoid overflow.

Github-Pull; #7480
Rebased-From: 993d089e82fc045d7b0f23e1a5dc934cba0e3306

sdaftuar committed:
Update nQueuedValidatedHeaders after peer disconnection

Github-Pull: #7482
Rebased-From: 301bc7bc7e83f4c268c1722558b07dbb5b55fa92

luke-jr committed:
release-notes: Mention possibility of priority removal in 0.13, uncertainty of priority calculation being changed back, and request community input

laanwj committed:
doc: Update release notes for rc4 changes

laanwj committed:
qt: Translations update pre-rc4

laanwj committed:
wallet: Ignore MarkConflict if block hash is not known

If number of conflict confirms cannot be determined, this means
that the block is still unknown or not yet part of the main chain,
for example during a reindex. Do nothing in that case,
instead of crash with an assertion.

Fixes #7234.

Github-Pull: #7491
Rebased-From: 40e7b61835cbe5fd471d0b4b71972526bf0e523c

luke-jr committed:
release-notes: Remove suggestion to use 0.11

gmaxwell committed:
release-notes: Significantly rewrite priority transactions section

luke-jr committed:
release-notes: Minor corrections and clarifications re Priority

sipa committed:
Correctly report high-S violations

Github-Pull: #7500
Rebased-From: 9d95187d5ddee56b6dfb55985008bdf70aed31f2

sipa committed:
Update the wallet best block marker when pruning

Github-Pull: #7502
Rebased-From: e4eebb604e19f67b0c7a483b1ded1229d75ecdd3

laanwj committed:
doc: Release notes update pre-rc5

laanwj committed:
qt: Translation update pre-rc5

laanwj committed:
doc: fix author list in release notes

laanwj committed:
Merge #7346: 0.12 release notes: Mining Policy Changes

b460004 release-notes: Minor corrections and clarifications re Priority (Luke Dashjr)
3450f4c release-notes: Significantly rewrite priority transactions section (Gregory Maxwell)
d0dbb6d release-notes: Remove suggestion to use 0.11 (Luke Dashjr)
73a0375 release-notes: Mention possibility of priority removal in 0.13, uncertainty of priority calculation being changed back, and request community input (Luke Dashjr)
4b8d2bc release-notes: Cover priority changes correctly, removing mentions of possible futures (Luke Dashjr)

Commit:
Fix of semantic failure "meet pay"

"do not meet pay the minimum relay fee" ? I can understand English language quite well, but that I do not understand. So, if it's not an semantic nonsense, I would suggest to write it in more simple English.

laanwj committed:
Merge #7523: Fix of semantic failure "meet pay"

e473c2d Fix of semantic failure "meet pay" (wodry)

laanwj committed:
doc: Remove another duplicate author name from release notes

The list of contributors is automatically generated from git, so people
that use multiple author names will end up on the list multiple times.

fanquake committed:
Fix duplicate names in release notes

Fixes #7547

laanwj committed:
Merge #7548: Correct duplicate names in release notes

ea52530 Fix duplicate names in release notes (fanquake)

sipa committed:
[HACK] Very temporary fix for reindex failure

sipa committed:
Add rewind logic to deal with post-fork software updates
